clang/lib/ASTMatchers/ASTMatchFinder.cpp

```diff
@@ -1237,7 +1237,8 @@ class MatchASTVisitor : public RecursiveASTVisitor<MatchASTVisitor>,
           // Make sure we do not visit the same node twice.
           // Otherwise, we'll visit the common ancestors as often as there
           // are splits on the way down.
-          if (Visited.insert(Parent.getMemoizationData()).second)
+          if (Parent.getMemoizationData() == nullptr ||
+              Visited.insert(Parent.getMemoizationData()).second)
```

Comment:
Note: Avoid double calls to getMemoizationData (assign to some local variable and use or write), but still call to getMemoizationData should be in theory optimized.

Reply:
Since this is simply an if and taking an address, I assumed it would be optimized. But I applied the suggested change.